presto-bytecode/src/main/java/com/facebook/presto/bytecode/ClassInfoLoader.java

```diff
@@ -0,0 +1,149 @@
+import org.objectweb.asm.util.CheckClassAdapter;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.facebook.presto.bytecode.ParameterizedType.typeFromPathName;
+
+public class ClassInfoLoader
+{
+    public static ClassInfoLoader createClassInfoLoader(Iterable<ClassDefinition> classDefinitions, ClassLoader classLoader)
+    {
+        ImmutableMap.Builder<ParameterizedType, ClassNode> classNodes = ImmutableMap.builder();
+        for (ClassDefinition classDefinition : classDefinitions) {
+            ClassNode classNode = new ClassNode();
+            classDefinition.visit(classNode);
+            classNodes.put(classDefinition.getType(), classNode);
+        }
+        return new ClassInfoLoader(classNodes.build(), ImmutableMap.of(), classLoader, true);
+    }
+
+    private final Map<ParameterizedType, ClassNode> classNodes;
+    private final Map<ParameterizedType, byte[]> bytecodes;
+    private final ClassLoader classLoader;
+    private final Map<ParameterizedType, ClassInfo> classInfoCache = new HashMap<>();
+    private final boolean loadMethodNodes;
+
+    public ClassInfoLoader(Map<ParameterizedType, ClassNode> classNodes, Map<ParameterizedType, byte[]> bytecodes, ClassLoader classLoader, boolean loadMethodNodes)
+    {
+        this.classNodes = ImmutableMap.copyOf(classNodes);
+        this.bytecodes = ImmutableMap.copyOf(bytecodes);
+        this.classLoader = classLoader;
+        this.loadMethodNodes = loadMethodNodes;
+    }
+
+    public ClassInfo loadClassInfo(ParameterizedType type)
+    {
+        ClassInfo classInfo = classInfoCache.get(type);
+        if (classInfo == null) {
+            classInfo = readClassInfoQuick(type);
+            classInfoCache.put(type, classInfo);
+        }
+        return classInfo;
+    }
+
+    private ClassInfo readClassInfoQuick(ParameterizedType type)
+    {
+        // check for user supplied class node
+        ClassNode classNode = classNodes.get(type);
+        if (classNode != null) {
+            return new ClassInfo(this, classNode);
+        }
+
+        // check for user supplied byte code
+        ClassReader classReader;
+        byte[] bytecode = bytecodes.get(type);
+        if (bytecode != null) {
+            classReader = new ClassReader(bytecode);
+        }
+        else {
+            // load class file from class loader
+            String classFileName = type.getClassName() + ".class";
+            try (InputStream is = classLoader.getResourceAsStream(classFileName)) {
+                classReader = new ClassReader(is);
+            }
+            catch (IOException e) {
+                // check if class is already loaded
+                try {
+                    Class<?> aClass = classLoader.loadClass(type.getJavaClassName());
+                    return new ClassInfo(this, aClass);
+                }
+                catch (ClassNotFoundException e1) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)